core/src/main/scala/kafka/server/metadata/QuotaCache.scala

```diff
@@ -0,0 +1,296 @@
+          quotaEntitySet.remove(quotaEntity)
+          quotaEntitySet.isEmpty
+        case None => false
+      }
+      if (needsCleanup) {
+        quotaCacheIndex.remove(key)
+      }
+    } else {
+      quotaCacheIndex.getOrElseUpdate(key, mutable.HashSet.empty).add(quotaEntity)
+    }
+  }
+
+  /**
+   * Update the quota cache with the given entity and quota key/value. If remove is set, the value is ignore and
+   * the quota entry is removed for the given key. No validation on quota keys is performed here, it is assumed
+   * that the caller has already done this.
+   *
+   * @param entity    A quota entity, either a specific entity or the default entity for the given type(s)
+   * @param key       The quota key
+   * @param value     The quota value
+   * @param remove    True if we should remove the given quota key from the entity's quota cache
+   */
+  def updateQuotaCache(entity: QuotaEntity, key: String, value: Double, remove: Boolean): Unit = inWriteLock(lock) {
+    val quotaValues = quotaCache.getOrElseUpdate(entity, mutable.HashMap.empty)
+    val removeFromIndex = if (remove) {
+      quotaValues.remove(key)
+      if (quotaValues.isEmpty) {
+        quotaCache.remove(entity)
+        true
+      } else {
+        false
+      }
+    } else {
+      quotaValues.put(key, value)
+      false
+    }
+
+    // Update the appropriate indexes with the entity
+    val updateCacheIndexPartial: (QuotaCacheIndex, CacheIndexKey) => Unit = updateCacheIndex(entity, removeFromIndex)
+    entity match {
+      case UserEntity(user) =>
+        updateCacheIndexPartial(userEntityIndex, SpecificUser(user))
+      case DefaultUserEntity =>
+        updateCacheIndexPartial(userEntityIndex, DefaultUser)
+
+      case ClientIdEntity(clientId) =>
+        updateCacheIndexPartial(clientIdEntityIndex, SpecificClientId(clientId))
+      case DefaultClientIdEntity =>
+        updateCacheIndexPartial(clientIdEntityIndex, DefaultClientId)
+
+      case ExplicitUserExplicitClientIdEntity(user, clientId) =>
+        updateCacheIndexPartial(userEntityIndex, SpecificUser(user))
+        updateCacheIndexPartial(clientIdEntityIndex, SpecificClientId(clientId))
+
+      case ExplicitUserDefaultClientIdEntity(user) =>
+        updateCacheIndexPartial(userEntityIndex, SpecificUser(user))
+        updateCacheIndexPartial(clientIdEntityIndex, DefaultClientId)
+
+      case DefaultUserExplicitClientIdEntity(clientId) =>
+        updateCacheIndexPartial(userEntityIndex, DefaultUser)
+        updateCacheIndexPartial(clientIdEntityIndex, SpecificClientId(clientId))
+
+      case DefaultUserDefaultClientIdEntity =>
+        updateCacheIndexPartial(userEntityIndex, DefaultUser)
+        updateCacheIndexPartial(clientIdEntityIndex, DefaultClientId)
+
+      case IpEntity(ip) =>
+        updateCacheIndexPartial(ipEntityIndex, SpecificIp(ip))
+      case DefaultIpEntity =>
+        updateCacheIndexPartial(ipEntityIndex, DefaultIp)
+    }
+  }
```

Comment:
Hmm. I think you can simplify this code if you instead have a function `convert` that knows how to convert a `QuotaEntity` to a `List[(QuotaCacheIndex, CacheIndexKey)]`. With this you can do something like:
```scala
convert(entity).foreach { (quotaCacheIndex, cacheIndexKey) =>
  updateCacheIndex(entity, removeFromIndex, quotaCacheIndex, cacheIndexKey)
}
```